server/src/main/java/io/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -157,53 +160,78 @@ public void stop()
     }
   }
 
+  private <T> T inReadOnlyTransaction(final TransactionCallback<T> callback)
+  {
+    return connector.getDBI().withHandle(
+        new HandleCallback<T>()
+        {
+          @Override
+          public T withHandle(Handle handle) throws Exception
+          {
+            final Connection connection = handle.getConnection();
+            final boolean readOnly = connection.isReadOnly();
+            connection.setReadOnly(true);
+            T res = handle.inTransaction(callback);
+            connection.setReadOnly(readOnly);
```

Comment:
`inTransaction` can throw an exception if it fails, do you know if that will cause the handle to get thrown away? If not then it would get stuck in read-only mode.